rclrs/src/lib.rs

```diff
@@ -53,16 +53,22 @@ pub fn spin_once(node: &Node, timeout: Option<Duration>) -> Result<(), RclrsErro
         &ctx,
     )?;
 
-    for live_subscription in &live_subscriptions {
-        wait_set.add_subscription(live_subscription.clone())?;
```

Comment:
Given that we the `Waitable`s are now typed, we can keep the old style (`Waitset::add_X(X)`). IMHO, it reads more logically that we add a `Subscription` to a `Waitset`. I'd agree more with the new style if we would have an arbitrary set of entities to be added to a `Waitset` (the responsibility of adding an entity should therefore be in the `Waitable` trait), but in this case, we know we won't have anything else other than `Subscription`s, `Client`s, `Service`s and `Timer`s, so we can just have methods for each entity in `Waitset`

Reply:
Ok, we need to keep the actual logic in `Waitable::add_to_wait_set` (see PR description for why), but I added back the old functions as syntactic sugar.

Reply:
I've read the description again, but I'm afraid I'm not following, sorry. Is it because you don't want to give `pub(crate)` visibility to `rcl_wait_set` in the `Waitset`? Is there a reason for that? To me `Waitset::add_subscription` seems the natural "translattion" for `rcl_wait_set_add_subscription`, so I would expect the logic for adding a subscription to be in `Waitset`.

Reply:
Hmm, no, I'm fine with `pub(crate)` visibility for rcl types. But the issue is that `WaitSet::add_subscription()` needs an `rcl_subscription_t`.

Prior to this PR, the situation is as follows: `WaitSet::add_subscription()` takes the subscription in the form of a `dyn SubscriptionBase`, and therefore that trait must be able to provide the `rcl_subscription_t`. We don't want to have the `rcl_subscription_t` in the public API, so `SubscriptionBase` returned the `rcl_subscription_t` in an opaque wrapper type, the `SubscriptionHandle`.

The main goal of this PR is to remove the `SubscriptionHandle`, since even though we now don't have the `rcl_subscription_t` in the public API, we now have that `SubscriptionHandle` in the public API which is also "internal" and has no useful functionality for the user. Fewer public types are better, as I'm sure you agree. It does that by changing the `handle()` method in `SubscriptionBase` to not *return* the `rcl_subscription_t` handle, but just *call the function that needs it* directly, in this case `rcl_wait_set_add_subscription()`. (To have a more fitting name, that method was then renamed to `add_to_wait_set()`, and the trait generalized into `Waitable`.)

So, I hope that makes sense – in Rust you cannot have a `pub(crate)` method on a trait, so all trait functions are public, so the way that the `SubscriptionBase` trait gives the `rcl_subscription_t` to the `WaitSet` is also public, unless we use the design in this PR.

We could maybe make the organization feel more "natural" like you say by moving the impls of `Waitable` into `wait.rs`. That way, the only place we deal with `rcl_wait_set_*` functions is in the `wait` module.

A third design would be to make `WaitSet::add_subscription()` take an actual `Subscription<T>`. Then we can just access the private parts, i.e. `rcl_subscription_t`. This would mean the `add_to_wait_set()` function is removed. It has the drawback of causing a bit more monomorphization, and being more restrictive on user code, since they'll not be able to e.g. downcast several subscriptions into a `SubscriptionBase` (or `SubscriptionWaitable`, if you're fine with that) and store them into a `Vec`, before adding them to a wait set.

Reply:
A fourth way is to keep `Handle` structs, but mark them as `#[doc(hidden)]`.

Reply:
I'd go with the third option, we know that `Waitset`s will only deal with `Subscription`s, `Client`s, `Service`s and `Timer`s, so I think it's fine to be that restrictive.

Otherwise, perhaps keeping the `Handle` structs is fine, a bit cumbersome, but they won't expose raw `rcl` types.

Reply:
> I'd go with the third option, we know that `Waitset`s will only deal with `Subscription`s, `Client`s, `Service`s and `Timer`s, so I think it's fine to be that restrictive.

Actually, sorry I only realize this now, but I don't think option 3 would work. The node needs to store the subscriptions (and clients and services), and it must do so by storing trait objects, not the fully-typed `Subscription<T>`, or `T` would need to become a type parameter of `Node` itself, which is impossible since there may be many subscriptions with different `T`s. And if it doesn't have the concrete type anymore, it cannot call a `WaitSet::add_subscription()` that takes the concrete `Subscription<T>` type.

So my opinion is that the design in this PR is the best one I can think of. It  The way things are currently done + `#[doc(hidden)]` is also acceptable – but if you don't like the design in this PR, I would like to learn more about why.

Reply:
> but if you don't like the design in this PR, I would like to learn more about why.

Not that I actively dislike it, but I find it unnecessarily complicated. The responsibility of adding a subscription to a waitset should belong in a waitset IMHO, as it's the one that's storing the subscription's handle. But with this PR, in order to allow the entities to be added to a waitset, the waitset handle needs to be made public, which does not seem to me necessary.

Reply:
If you mean the `rcl_wait_set_t` by "waitset handle", that's not public. The trait uses an `rclrs::WaitSet`.

Reply:
Then let's leave this PR aside for a while and let me rework #224 to work with `SubscriptionHandle`. Then we can release and later worry about removing `SubscriptionHandle`.